File: package/main/src/Predicate/matches.ts
/**
 * Creates a predicate that checks whether an object matches
 * all properties of the given pattern using strict equality
 * @param {Record<string, unknown>} pattern - The pattern to match against
 * @returns {(obj: Record<string, unknown>) => boolean} A predicate that tests objects
 * @example
 * const isAdmin = matches({ role: "admin" });
 * isAdmin({ name: "Alice", role: "admin" }); // true
 * isAdmin({ name: "Bob", role: "user" }); // false
 */
export const matches =
  (
    pattern: Record<string, unknown>,
  ): ((object: Record<string, unknown>) => boolean) =>
  (object: Record<string, unknown>): boolean => {
    const keys = Object.keys(pattern);
    for (const key of keys) {
      if (object[key] !== pattern[key]) {
        return false;
      }
    }
    return true;
  };

File: package/umt_python/src/predicate/matches.py
from collections.abc import Callable
from typing import Any


def matches(pattern: dict[str, Any]) -> Callable[[dict[str, Any]], bool]:
    """
    Creates a predicate that checks whether an object matches
    all properties of the given pattern using strict equality

    Args:
        pattern: The pattern to match against

    Returns:
        A predicate that tests objects

    Example:
        >>> is_admin = matches({"role": "admin"})
        >>> is_admin({"name": "Alice", "role": "admin"})
        True
        >>> is_admin({"name": "Bob", "role": "user"})
        False
    """

    def matcher(obj: dict[str, Any]) -> bool:
        for key, value in pattern.items():
            if key not in obj:
                return False
            obj_val = obj[key]
            if type(obj_val) is bool and type(value) is not bool:
                return False
            if type(obj_val) is not bool and type(value) is bool:
                return False
            if obj_val != value:
                return False
        return True

    return matcher
